File: src/Message.zig
const std = @import("std");
const df = @import("ImportC.zig").df;
const Window = @import("Window.zig");

const MAXMESSAGES = 100;

// ---------- event queue ----------
const Evt = struct {
    event:df.MESSAGE,
    mx:c_int,
    my:c_int,
};

var EventQueue = [_]Evt{.{.event=0, .mx=0, .my=0}}**MAXMESSAGES;

var EventQueueOnCtr:usize = 0;
var EventQueueOffCtr:usize = 0;
var EventQueueCtr:usize = 0;

// ---------- message queue ---------
const Msg = struct {
    wnd:df.WINDOW,
    msg:df.MESSAGE,
    p1:df.PARAM,
    p2:df.PARAM,
};

var MsgQueue = [_]Msg{.{.wnd=null, .msg=0, .p1=0, .p2=0}}**MAXMESSAGES;

var MsgQueueOnCtr:usize = 0;
var MsgQueueOffCtr:usize = 0;
var MsgQueueCtr:usize = 0;

// ----- post an event and parameters to event queue ----
pub export fn PostEvent(event:df.MESSAGE, p1:c_int, p2:c_int) callconv(.c) void {
    if (EventQueueCtr != MAXMESSAGES) {
        EventQueue[EventQueueOnCtr].event = event;
        EventQueue[EventQueueOnCtr].mx = p1;
        EventQueue[EventQueueOnCtr].my = p2;
        EventQueueOnCtr += 1;
        if (EventQueueOnCtr == MAXMESSAGES) {
            EventQueueOnCtr = 0;
        }
        EventQueueCtr += 1;
    }
}

// ----- post a message and parameters to msg queue ----
pub export fn PostMessage(wnd:df.WINDOW, msg:df.MESSAGE, p1:df.PARAM, p2:df.PARAM) callconv(.c) void {
    if (MsgQueueCtr != MAXMESSAGES) {
        MsgQueue[MsgQueueOnCtr].wnd = wnd;
        MsgQueue[MsgQueueOnCtr].msg = msg;
        MsgQueue[MsgQueueOnCtr].p1 = p1;
        MsgQueue[MsgQueueOnCtr].p2 = p2;
        MsgQueueOnCtr += 1;
        if (MsgQueueOnCtr == MAXMESSAGES) {
            MsgQueueOnCtr = 0;
        }
        MsgQueueCtr += 1;
    }
}

// ------------ initialize the message system ---------
pub fn init_messages() bool {
    var cols:c_int = 0;
    var rows:c_int = 0;

    df.AllocTesting = df.TRUE;
    if (df.setjmp(&df.AllocError) != 0) {
        df.StopMsg();
        return false;
    }

    _ = df.tty_init(df.MouseTracking|df.CatchISig|df.ExitLastLine|df.FullBuffer);
    if (df.tty_getsize(&cols, &rows) > 0) {
        df.SCREENWIDTH = @min(cols, df.MAXCOLS-1);
        df.SCREENHEIGHT = rows - 1;
    }

    df.resetmouse();
    df.set_mousetravel(0, df.SCREENWIDTH-1, 0, df.SCREENHEIGHT-1);
    df.savecursor();
    df.hidecursor();

    df.CaptureMouse = null;
    df.CaptureKeyboard = null;

    _ = df.init_messages();
    return true;
}

// ---- dispatch messages to the message proc function ----
pub fn dispatch_message() bool {
    // -------- collect mouse and keyboard events -------
    df.collect_events();

    // --------- dequeue and process events --------
    while (EventQueueCtr > 0)  {
        const ev = EventQueue[EventQueueOffCtr];
        EventQueueOffCtr += 1;
        if (EventQueueOffCtr == MAXMESSAGES)
            EventQueueOffCtr = 0;
        EventQueueCtr -= 1;

        df.cdispatch_message(ev.event, ev.mx, ev.my);
    }

    // ------ dequeue and process messages -----
    while (MsgQueueCtr > 0) {
        const mq = MsgQueue[MsgQueueOffCtr];
        MsgQueueOffCtr += 1;
        if (MsgQueueOffCtr == MAXMESSAGES) {
            MsgQueueOffCtr = 0;
        }
        MsgQueueCtr -= 1;
        _ = df.SendMessage(mq.wnd, mq.msg, mq.p1, mq.p2);
        if (mq.msg == df.ENDDIALOG) {
            return false;
        }
        if (mq.msg == df.STOP) {
            _ = df.PostMessage(null, df.STOP, 0, 0);
            return false;
        }
    }

    // #define VIDEO_FB 1
    df.convert_screen_to_ansi();

    return true;
}

// ----------- dflatmsg.h ------------

// message foundation file
// make message changes here
// other source files will adapt

pub const Message = enum (c_int) {
    // -------------- process communication messages -----------
    START = df.START,                    // start message processing
    STOP = df.STOP,                      // stop message processing
    COMMAND = df.COMMAND,                // send a command to a window
    // -------------- window management messages ---------------
    CREATE_WINDOW = df.CREATE_WINDOW,    // create a window
    OPEN_WINDOW = df.OPEN_WINDOW,        // open a window
    SHOW_WINDOW = df.SHOW_WINDOW,        // show a window
    HIDE_WINDOW = df.HIDE_WINDOW,        // hide a window
    CLOSE_WINDOW = df.CLOSE_WINDOW,      // delete a window
    SETFOCUS = df.SETFOCUS,              // set and clear the focus
    PAINT = df.PAINT,                    // paint the window's data space
    BORDER = df.BORDER,                  // paint the window's border
    TITLE = df.TITLE,                    // display the window's title
    MOVE = df.MOVE,                      // move the window
    SIZE = df.SIZE,                      // change the window's size
    MAXIMIZE = df.MAXIMIZE,              // maximize the window
    MINIMIZE = df.MINIMIZE,              // minimize the window
    RESTORE = df.RESTORE,                // restore the window
    INSIDE_WINDOW = df.INSIDE_WINDOW,    // test x/y inside a window
    // ---------------- clock messages -------------------------
    CLOCKTICK = df.CLOCKTICK,            // the clock ticked
    CAPTURE_CLOCK = df.CAPTURE_CLOCK,    // capture clock into a window
    RELEASE_CLOCK = df.RELEASE_CLOCK,    // release clock to the system
    // -------------- keyboard and screen messages -------------
    KEYBOARD = df.KEYBOARD,                   // key was pressed
    CAPTURE_KEYBOARD = df.CAPTURE_KEYBOARD,   // capture keyboard into a window
    RELEASE_KEYBOARD = df.RELEASE_KEYBOARD,   // release keyboard to system
    KEYBOARD_CURSOR = df.KEYBOARD_CURSOR,     // position the keyboard cursor
    CURRENT_KEYBOARD_CURSOR = df.CURRENT_KEYBOARD_CURSOR, //read the cursor position
    HIDE_CURSOR = df.HIDE_CURSOR,             // hide the keyboard cursor
    SHOW_CURSOR = df.SHOW_CURSOR,             // display the keyboard cursor
    SAVE_CURSOR = df.SAVE_CURSOR,             // save the cursor's configuration
    RESTORE_CURSOR = df.RESTORE_CURSOR,       // restore the saved cursor
    SHIFT_CHANGED = df.SHIFT_CHANGED,         // the shift status changed
    WAITKEYBOARD = df.WAITKEYBOARD,           // waits for a key to be released
    // ---------------- mouse messages -------------------------
    RESET_MOUSE = df.RESET_MOUSE,             // reset the mouse
    MOUSE_TRAVEL = df.MOUSE_TRAVEL,           // set the mouse travel
    MOUSE_INSTALLED = df.MOUSE_INSTALLED,     // test for mouse installed
    RIGHT_BUTTON = df.RIGHT_BUTTON,           // right button pressed
    LEFT_BUTTON = df.LEFT_BUTTON,             // left button pressed
    DOUBLE_CLICK = df.DOUBLE_CLICK,           // left button double-clicked
    MOUSE_MOVED = df.MOUSE_MOVED,             // mouse changed position
    BUTTON_RELEASED = df.BUTTON_RELEASED,     // mouse button released
    CURRENT_MOUSE_CURSOR = df.CURRENT_MOUSE_CURSOR, // get mouse position
    MOUSE_CURSOR = df.MOUSE_CURSOR,           // set mouse position
    SHOW_MOUSE = df.SHOW_MOUSE,               // make mouse cursor visible
    HIDE_MOUSE = df.HIDE_MOUSE,               // hide mouse cursor
    WAITMOUSE = df.WAITMOUSE,                 // wait until button released
    TESTMOUSE = df.TESTMOUSE,                 // test any mouse button pressed
    CAPTURE_MOUSE = df.CAPTURE_MOUSE,         // capture mouse into a window
    RELEASE_MOUSE = df.RELEASE_MOUSE,         // release the mouse to system
    // ---------------- text box messages ----------------------
    ADDTEXT = df.ADDTEXT,                     // append text to the text box
    INSERTTEXT = df.INSERTTEXT,               // insert line of text
    DELETETEXT = df.DELETETEXT,               // delete line of text
    CLEARTEXT = df.CLEARTEXT,                 // clear the edit box
    SETTEXT = df.SETTEXT,                     // copy text to text buffer
    SCROLL = df.SCROLL,                       // vertical line scroll
    HORIZSCROLL = df.HORIZSCROLL,             // horizontal column scroll
    SCROLLPAGE = df.SCROLLPAGE,               // vertical page scroll
    HORIZPAGE = df.HORIZPAGE,                 // horizontal page scroll
    SCROLLDOC = df.SCROLLDOC,                 // scroll to beginning/end
    // ---------------- edit box messages ----------------------
    GETTEXT = df.GETTEXT,                     // get text from an edit box
    SETTEXTLENGTH = df.SETTEXTLENGTH,         // set maximum text length
    // ---------------- menubar messages -----------------------
    BUILDMENU = df.BUILDMENU,                 // build the menu display
    MB_SELECTION = df.MB_SELECTION,           // menubar selection
    // ---------------- popdown messages -----------------------
    BUILD_SELECTIONS = df.BUILD_SELECTIONS,   // build the menu display
    CLOSE_POPDOWN = df.CLOSE_POPDOWN,         // tell parent popdown is closing
    // ---------------- list box messages ----------------------
    LB_SELECTION = df.LB_SELECTION,           // sent to parent on selection
    LB_CHOOSE = df.LB_CHOOSE,                 // sent when user chooses
    LB_CURRENTSELECTION = df.LB_CURRENTSELECTION, // return the current selection
    LB_GETTEXT = df.LB_GETTEXT,               // return the text of selection
    LB_SETSELECTION = df.LB_SETSELECTION,     // sets the listbox selection
    // ---------------- dialog box messages --------------------
    INITIATE_DIALOG = df.INITIATE_DIALOG,     // begin a dialog
    ENTERFOCUS = df.ENTERFOCUS,               // tell DB control got focus
    LEAVEFOCUS = df.LEAVEFOCUS,               // tell DB control lost focus
    ENDDIALOG = df.ENDDIALOG,                 // end a dialog
    // ---------------- help box messages ----------------------
    DISPLAY_HELP = df.DISPLAY_HELP,
    // --------------- application window messages -------------
    ADDSTATUS = df.ADDSTATUS,
    // --------------- picture box messages --------------------
    DRAWVECTOR = df.DRAWVECTOR,
    DRAWBOX = df.DRAWBOX,
    DRAWBAR = df.DRAWBAR,
};

File: message.c
/* --------- message.c ---------- */

#include "dflat.h"

//static int px = -1, py = -1;
//static int pmx = -1, pmy = -1;
//static int mx, my;
static int handshaking = 0;
BOOL AllocTesting = FALSE;
jmp_buf AllocError;
BOOL AltDown = FALSE;

/* ---------- event queue ---------- */
static struct events    {
    MESSAGE event;
    int mx;
    int my;
} EventQueue[MAXMESSAGES];

/* ---------- message queue --------- */
static struct msgs {
    WINDOW wnd;
    MESSAGE msg;
    PARAM p1;
    PARAM p2;
} MsgQueue[MAXMESSAGES];

static int EventQueueOnCtr;
static int EventQueueOffCtr;
static int EventQueueCtr;

static int MsgQueueOnCtr;
static int MsgQueueOffCtr;
static int MsgQueueCtr;

static int lagdelay = FIRSTDELAY;

static volatile int keyportvalue;	/* for watching for key release */

WINDOW CaptureMouse;
WINDOW CaptureKeyboard;
static BOOL NoChildCaptureMouse;
static BOOL NoChildCaptureKeyboard;

static volatile int delaytimer  = -1;
static volatile int clocktimer  = -1;
char time_string[] = "         ";

static WINDOW Cwnd;

static void StopMsg(void)
{
#ifndef BUILD_SMALL_DFLAT
	ClearClipboard();
	ClearDialogBoxes();
#endif
	restorecursor();	
	unhidecursor();
    hide_mousecursor();
}

int ProcessMessage(WINDOW wnd, MESSAGE msg, PARAM p1, PARAM p2);

/* ------------ initialize the message system --------- */
BOOL init_messages(void)
{
    int cols, rows;

	AllocTesting = TRUE;
	if (setjmp(AllocError) != 0)	{
		StopMsg();
		return FALSE;
	}
    tty_init(MouseTracking|CatchISig|ExitLastLine|FullBuffer);
    if (tty_getsize(&cols, &rows) > 0) {
        SCREENWIDTH = min(cols, MAXCOLS-1);
        SCREENHEIGHT = rows - 1;
    }
    resetmouse();
	set_mousetravel(0, SCREENWIDTH-1, 0, SCREENHEIGHT-1);
	savecursor();
	hidecursor();
//    px = py = -1;
//    pmx = pmy = -1;
//    mx = my = 0;
    CaptureMouse = CaptureKeyboard = NULL;
    NoChildCaptureMouse = FALSE;
    NoChildCaptureKeyboard = FALSE;
    MsgQueueOnCtr = MsgQueueOffCtr = MsgQueueCtr = 0;
    EventQueueOnCtr = EventQueueOffCtr = EventQueueCtr = 0;
    PostMessage(NULL,START,0,0);
    lagdelay = FIRSTDELAY;
	return TRUE;
}

/* ----- post an event and parameters to event queue ---- */
void PostEvent(MESSAGE event, int p1, int p2)
{
    if (EventQueueCtr != MAXMESSAGES)    {
        EventQueue[EventQueueOnCtr].event = event;
        EventQueue[EventQueueOnCtr].mx = p1;
        EventQueue[EventQueueOnCtr].my = p2;
        if (++EventQueueOnCtr == MAXMESSAGES)
            EventQueueOnCtr = 0;
        EventQueueCtr++;
    }
}

/* ----- post a message and parameters to msg queue ---- */
void PostMessage(WINDOW wnd, MESSAGE msg, PARAM p1, PARAM p2)
{
    if (MsgQueueCtr != MAXMESSAGES)    {
        MsgQueue[MsgQueueOnCtr].wnd = wnd;
        MsgQueue[MsgQueueOnCtr].msg = msg;
        MsgQueue[MsgQueueOnCtr].p1 = p1;
        MsgQueue[MsgQueueOnCtr].p2 = p2;
        if (++MsgQueueOnCtr == MAXMESSAGES)
            MsgQueueOnCtr = 0;
        MsgQueueCtr++;
    }
}

/* --------- send a message to a window ----------- */
int SendMessage(WINDOW wnd, MESSAGE msg, PARAM p1, PARAM p2)
{
    int rtn = TRUE, x, y;

#ifdef INCLUDE_LOGGING
	LogMessages(wnd, msg, p1, p2);
#endif
    if (wnd != NULL) {
        switch (msg)    {
            case PAINT:
            case BORDER:
                /* ------- don't send these messages unless the
                    window is visible -------- */
                if (isVisible(wnd))
	                rtn = (*wnd->wndproc)(wnd, msg, p1, p2);
                break;
            case RIGHT_BUTTON:
            case LEFT_BUTTON:
            case DOUBLE_CLICK:
            case BUTTON_RELEASED:
                /* --- don't send these messages unless the
                    window is visible or has captured the mouse -- */
                if (isVisible(wnd) || wnd == CaptureMouse)
	                rtn = (*wnd->wndproc)(wnd, msg, p1, p2);
                break;
            case KEYBOARD:
            case SHIFT_CHANGED:
                /* ------- don't send these messages unless the
                    window is visible or has captured the keyboard -- */
                if (!(isVisible(wnd) || wnd == CaptureKeyboard))
	                break;
            default:
                rtn = (*wnd->wndproc)(wnd, msg, p1, p2);
                break;
        }
    }
    /* ----- window processor returned true or the message was sent
        to no window at all (NULL) ----- */
    if (rtn != FALSE)    {
        rtn = ProcessMessage(wnd, msg, p1, p2);
    }
    return rtn;
}

// This one follows SendMessage();
int ProcessMessage(WINDOW wnd, MESSAGE msg, PARAM p1, PARAM p2)
{
    int rtn = TRUE, x, y;
        /* --------- process messages that a window sends to the
            system itself ---------- */
        switch (msg)    {
            case STOP:
				StopMsg();
                break;
            /* ------- clock messages --------- */
            case CAPTURE_CLOCK:
				if (Cwnd == NULL)
	                set_timer(clocktimer, 0);
				wnd->PrevClock = Cwnd;
                Cwnd = wnd;
                break;
            case RELEASE_CLOCK:
                Cwnd = wnd->PrevClock;
				if (Cwnd == NULL)
	                disable_timer(clocktimer);
                break;
            /* -------- keyboard messages ------- */
            case KEYBOARD_CURSOR:
                if (wnd == NULL)
                    cursor((int)p1, (int)p2);
                else if (wnd == inFocus)
                    cursor(GetClientLeft(wnd)+(int)p1,
                                GetClientTop(wnd)+(int)p2);
                break;
            case CAPTURE_KEYBOARD:
                if (p2)
                    ((WINDOW)p2)->PrevKeyboard=CaptureKeyboard;
                else
                    wnd->PrevKeyboard = CaptureKeyboard;
                CaptureKeyboard = wnd;
                NoChildCaptureKeyboard = (int)p1;
                break;
            case RELEASE_KEYBOARD:
				if (wnd != NULL)	{
					if (CaptureKeyboard == wnd || (int)p1)
	                	CaptureKeyboard = wnd->PrevKeyboard;
					else	{
						WINDOW twnd = CaptureKeyboard;
						while (twnd != NULL)	{
							if (twnd->PrevKeyboard == wnd)	{
								twnd->PrevKeyboard = wnd->PrevKeyboard;
								break;
							}
							twnd = twnd->PrevKeyboard;
						}
						if (twnd == NULL)
							CaptureKeyboard = NULL;
					}
                	wnd->PrevKeyboard = NULL;
				}
				else
					CaptureKeyboard = NULL;
                NoChildCaptureKeyboard = FALSE;
                break;
            case CURRENT_KEYBOARD_CURSOR:
                curr_cursor(&x, &y);
                *(int*)p1 = x;
                *(int*)p2 = y;
                break;
            case SAVE_CURSOR:
                savecursor();
                break;
            case RESTORE_CURSOR:
                restorecursor();
                break;
            case HIDE_CURSOR:
                normalcursor();
                hidecursor();
                break;
            case SHOW_CURSOR:
                if (p1)
                    set_cursor_type(0x0106);
                else
                    set_cursor_type(0x0607);
                unhidecursor();
                break;
			case WAITKEYBOARD:
				waitforkeyboard();
				break;
            /* -------- mouse messages -------- */
			case RESET_MOUSE:
				resetmouse();
				set_mousetravel(0, SCREENWIDTH-1, 0, SCREENHEIGHT-1);
				break;
            case MOUSE_INSTALLED:
                rtn = mouse_installed();
                break;
			case MOUSE_TRAVEL:	{
				RECT rc;
				if (!p1)	{
        			rc.lf = rc.tp = 0;
        			rc.rt = SCREENWIDTH-1;
        			rc.bt = SCREENHEIGHT-1;
				}
				else 
					rc = *(RECT *)p1;
				set_mousetravel(rc.lf, rc.rt, rc.tp, rc.bt);
				break;
			}
            case SHOW_MOUSE:
                show_mousecursor();
                break;
            case HIDE_MOUSE:
                hide_mousecursor();
                break;
            case MOUSE_CURSOR:
                set_mouseposition((int)p1, (int)p2);
                break;
            case CURRENT_MOUSE_CURSOR:
                get_mouseposition((int*)p1,(int*)p2);
                break;
            case WAITMOUSE:
                waitformouse();
                break;
            case TESTMOUSE:
                rtn = mousebuttons();
                break;
            case CAPTURE_MOUSE:
                if (p2)
                    ((WINDOW)p2)->PrevMouse = CaptureMouse;
                else
                    wnd->PrevMouse = CaptureMouse;
                CaptureMouse = wnd;
                NoChildCaptureMouse = (int)p1;
                break;
            case RELEASE_MOUSE:
				if (wnd != NULL)	{
					if (CaptureMouse == wnd || (int)p1)
	                	CaptureMouse = wnd->PrevMouse;
					else	{
						WINDOW twnd = CaptureMouse;
						while (twnd != NULL)	{
							if (twnd->PrevMouse == wnd)	{
								twnd->PrevMouse = wnd->PrevMouse;
								break;
							}
							twnd = twnd->PrevMouse;
						}
						if (twnd == NULL)
							CaptureMouse = NULL;
					}
                	wnd->PrevMouse = NULL;
				}
				else
					CaptureMouse = NULL;
                NoChildCaptureMouse = FALSE;
                break;
            default:
                break;
        }
    return rtn;
}

static RECT VisibleRect(WINDOW wnd)
{
	RECT rc = WindowRect(wnd);
	if (!TestAttribute(wnd, NOCLIP))	{
		WINDOW pwnd = GetParent(wnd);
		if (!pwnd)
			return rc;
		RECT prc;
		prc = ClientRect(pwnd);
		while (pwnd != NULL)	{
			if (TestAttribute(pwnd, NOCLIP))
				break;
			rc = subRectangle(rc, prc);
			if (!ValidRect(rc))
				break;
			if ((pwnd = GetParent(pwnd)) != NULL)
				prc = ClientRect(pwnd);
		}
	}
	return rc;
}

/* ----- find window that mouse coordinates are in --- */
static WINDOW inWindow(WINDOW wnd, int x, int y)
{
	WINDOW Hit = NULL;
	while (wnd != NULL)	{
		if (isVisible(wnd))	{
			WINDOW wnd1;
			RECT rc = VisibleRect(wnd);
			if (InsideRect(x, y, rc))
				Hit = wnd;
			if ((wnd1 = inWindow(LastWindow(wnd), x, y)) != NULL)
				Hit = wnd1;
			if (Hit != NULL)
				break;
		}
		wnd = PrevWindow(wnd);
	}
	return Hit;
}

static WINDOW MouseWindow(int x, int y)
{
    /* ------ get the window in which a
                    mouse event occurred ------ */
    WINDOW Mwnd = inWindow(ApplicationWindow, x, y);
    /* ---- process mouse captures ----- */
    if (CaptureMouse != NULL)	{
        if (NoChildCaptureMouse ||
				Mwnd == NULL 	||
					!isAncestor(Mwnd, CaptureMouse))
            Mwnd = CaptureMouse;
	}
	return Mwnd;
}

void handshake(void)
{
#if MSDOS
	/*
	handshaking++;
	dispatch_message();
	--handshaking;
	*/
#endif
}

/* ---- dispatch messages to the message proc function ---- */
BOOL dispatch_message(void)
{
    WINDOW Mwnd, Kwnd;
    /* -------- collect mouse and keyboard events ------- */
    collect_events();
    /* --------- dequeue and process events -------- */
    while (EventQueueCtr > 0)  {
        struct events ev;
			
		ev = EventQueue[EventQueueOffCtr];
        if (++EventQueueOffCtr == MAXMESSAGES)
            EventQueueOffCtr = 0;
        --EventQueueCtr;

        /* ------ get the window in which a
                        keyboard event occurred ------ */
        Kwnd = inFocus;

        /* ---- process keyboard captures ----- */
        if (CaptureKeyboard != NULL)
            if (Kwnd == NULL ||
                    NoChildCaptureKeyboard ||
						!isAncestor(Kwnd, CaptureKeyboard))
                Kwnd = CaptureKeyboard;

        /* -------- send mouse and keyboard messages to the
            window that should get them -------- */
        switch (ev.event)    {
            case SHIFT_CHANGED:
            case KEYBOARD:
				if (!handshaking)
	                SendMessage(Kwnd, ev.event, ev.mx, ev.my);
                break;
            case LEFT_BUTTON:
				if (!handshaking)	{
		        	Mwnd = MouseWindow(ev.mx, ev.my);
                	if (!CaptureMouse ||
                        	(!NoChildCaptureMouse &&
								isAncestor(Mwnd, CaptureMouse)))
                    	if (Mwnd != inFocus)
                        	SendMessage(Mwnd, SETFOCUS, TRUE, 0);
                	SendMessage(Mwnd, LEFT_BUTTON, ev.mx, ev.my);
				}
                break;
            case BUTTON_RELEASED:
            case DOUBLE_CLICK:
            case RIGHT_BUTTON:
				if (handshaking)
					break;
            case MOUSE_MOVED:
		        Mwnd = MouseWindow(ev.mx, ev.my);
                SendMessage(Mwnd, ev.event, ev.mx, ev.my);
                break;
#if MSDOS	// FIXME add MK_FP
            case CLOCKTICK:
                SendMessage(Cwnd, ev.event,
                    (PARAM) MK_FP(ev.mx, ev.my), 0);
				break;
#endif
            default:
                break;
        }
    }
    /* ------ dequeue and process messages ----- */
    while (MsgQueueCtr > 0)  {
        struct msgs mq;

		mq = MsgQueue[MsgQueueOffCtr];
        if (++MsgQueueOffCtr == MAXMESSAGES)
            MsgQueueOffCtr = 0;
        --MsgQueueCtr;
        SendMessage(mq.wnd, mq.msg, mq.p1, mq.p2);
        if (mq.msg == ENDDIALOG)
			return FALSE;
        if (mq.msg == STOP)	{
		    PostMessage(NULL, STOP, 0, 0);
			return FALSE;
		}
    }
#if VIDEO_FB
    convert_screen_to_ansi();
#endif
    return TRUE;
}
